src/covid19model/optimization/objective_fcns.py

```diff
@@ -1,5 +1,76 @@
 import numpy as np
 
+
+def sse(theta, model, data, model_parameters, lag_time=None):
+    """Calculate the sum of squared errors from data and model instance
+
+    The function assumes the number of values in the theta array corresponds
+    to the number of defined parameters, but can be extended with a lag_time
+    parameters to make this adjustable as well by an optimization algorithm.
+
+
+    Parameters
+    ----------
+    theta : np.array
+        Array with N (if lag_time is defined) or N+1 (if lag_time is not
+        defined) values.
+    model : covidmodel.model instance
+        Covid model instance
+    data : xarray.DataArray
+        Xarray DataArray with time as a dimension and the name corresponding
+        to an existing model output state.
+    model_parameters : list of str
+        Parameter names of the model parameters to adjust in order to get the
+        SSE.
+    lag_time : int
+        Warming up period before comparing the data with the model output.
+        e.g. if 40; comparison of data only starts after 40 days.
+
+
+    Notes
+    -----
+    Assumes daily time step(!) # TODO: need to generalize this
+
+    Examples
+    --------
+    >>> data_to_fit = xr.DataArray([ 54,  79, 100, 131, 165, 228, 290],
+                                   coords={"time": range(1, 8)}, name="ICU",
+                                   dims=['time'])
+    >>> sse([1.6, 0.025, 42], sir_model, data_to_fit, ['sigma', 'beta'], lag_time=None)
+    >>> # previous is equivalent to
+    >>> sse([1.6, 0.025], sir_model, data_to_fit, ['sigma', 'beta'], lag_time=42)
+    >>> # but the latter will use a fixed lag_time,
+    >>> # whereas the first one can be used inside optimization
+    """
+
+    if data.name not in model.state_names:
+        raise Exception("Data variable to fit is not available as model output")
+
+    # define new parameters in model
+    for i, parameter in enumerate(model_parameters):
+        model.parameters.update({parameter: theta[i]})
+
+    # extract additional parameter # TODO - check alternatives for generalisation here
+    if not lag_time:
+        lag_time = int(round(theta[-1]))
+    else:
+        lag_time = int(round(lag_time))
+
+    # run model
+    time = len(data.time) + lag_time # at least this length
+    output = model.sim(time)
```

Comment:
Could also run with `[-lag_time, time]` ? (or does the model not work with negative time?)

Reply:
The `solve_ivp` from SciPy requires the user to define the times at which the model output should be returned, they can be negative without any problem. So assigning the time as `[ -lag_time, time]`  in the model output xarray, and assigning the time as `[0 1 2 ... 7]` for the data should allow for pairing without using values as suggested below. I'm wondering if this will work if the times do not match, f.i. when using a discrete-time model with step-length 0.235, some form of extrapolation will be necessary. Xarray can be linked to SciPy interpolate: see http://xarray.pydata.org/en/stable/generated/xarray.DataArray.interp.html

Reply:
The default will be to only align those values where the time values match exactly, but you can indeed make this more flexible (eg http://xarray.pydata.org/en/stable/generated/xarray.DataArray.reindex.html also allows a simple "nearest" or a certain tolerance that should be allowed)